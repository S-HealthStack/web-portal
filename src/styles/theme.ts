import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
