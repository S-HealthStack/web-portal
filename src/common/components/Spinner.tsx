import React from 'react';

import styled, { keyframes, css } from 'styled-components';

import SpinnerIconXS from 'src/assets/icons/spinner_xs.svg';
import SpinnerIconM from 'src/assets/icons/spinner_m.svg';
import SpinnerIconL from 'src/assets/icons/spinner_l.svg';
import { colors, animation } from 'src/styles';

const spinnerAnimation = keyframes`
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
`;

type Size = 'xs' | 'm' | 'l';

type Props = {
  size: Size;
  $light?: boolean;
};

const spinnerStyles = ($light = false) => css`
  fill: ${$light ? colors.surface : colors.primary};
  animation: 1.1s infinite ${spinnerAnimation} ${animation.defaultTiming};
`;

const SpinnerIconXSStyled = styled(SpinnerIconXS)<{ $light?: boolean }>`
  ${({ $light }) => spinnerStyles($light)}
`;

const SpinnerIconMStyled = styled(SpinnerIconM)<{ $light?: boolean }>`
  ${({ $light }) => spinnerStyles($light)}
`;

const SpinnerIconLStyled = styled(SpinnerIconL)<{ $light?: boolean }>`
  ${({ $light }) => spinnerStyles($light)}
`;

const getSpinnerWithSize = (s: Size, $light = false) => {
  let icon = <SpinnerIconXSStyled $light={$light} data-testid="spinner" />;
  switch (s) {
    case 'm':
      icon = <SpinnerIconMStyled $light={$light} data-testid="spinner" />;
      break;
    case 'l':
      icon = <SpinnerIconLStyled $light={$light} data-testid="spinner" />;
      break;
    default:
      break;
  }
  return icon;
};

const Spinner = ({ size, $light }: Props) => getSpinnerWithSize(size, $light);

export default Spinner;
