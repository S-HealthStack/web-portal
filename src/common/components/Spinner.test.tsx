import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import theme from 'src/styles/theme';
import { ThemeProvider } from 'styled-components/';
import Spinner from './Spinner';

describe('Spinner', () => {
  it('test spinner render', () => {
    const { baseElement, queryByTestId, rerender } = render(
      <ThemeProvider theme={theme}>
        <Spinner size="xs" />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const spinner = queryByTestId('spinner') as Element;

    expect(spinner).toBeInTheDocument();
    expect(spinner).toHaveStyle(`fill: ${theme.colors.primary}`);

    rerender(
      <ThemeProvider theme={theme}>
        <Spinner size="xs" $light />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();
    expect(spinner).toBeInTheDocument();
    expect(spinner).toHaveStyle(`fill: ${theme.colors.surface}`);
  });

  it('[NEGATIVE] should render with wrong props', () => {
    const { baseElement, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Spinner size={'unknown' as 'xs'} />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const spinner = queryByTestId('spinner') as Element;

    expect(spinner).toBeInTheDocument();
  });

  it('[NEGATIVE] should render without props', () => {
    const { baseElement, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Spinner />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();
    expect(queryByTestId('spinner')).toBeInTheDocument();
  });
});
