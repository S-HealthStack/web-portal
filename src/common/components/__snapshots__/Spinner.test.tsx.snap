// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner test spinner render 1`] = `
.c0 {
  fill: #4475E3;
  -webkit-animation: 1.1s infinite cilQsd cubic-bezier(0.45,0.05,0,1);
  animation: 1.1s infinite cilQsd cubic-bezier(0.45,0.05,0,1);
}

<body>
  <div>
    <svg
      class="c0"
      data-testid="spinner"
    />
  </div>
</body>
`;

exports[`Spinner test spinner render 2`] = `
.c0 {
  fill: #FFFFFF;
  -webkit-animation: 1.1s infinite cilQsd cubic-bezier(0.45,0.05,0,1);
  animation: 1.1s infinite cilQsd cubic-bezier(0.45,0.05,0,1);
}

<body>
  <div>
    <svg
      class="c0"
      data-testid="spinner"
    />
  </div>
</body>
`;
